### ISSUE TITLE
Required fields being marked nullable

### ISSUE THREAD

**Issue by saward** [go-swagger/go-swagger#1525]
## Problem statement

I'm generating a client for the EVE Online ESI API, using https://esi.evetech.net/latest/swagger.json.

For the universe/types/get type ID endpoint, there are a handful of fields marked as required (https://esi.evetech.net/ui/#/Universe/get_universe_types_type_id):

```
"required": [
"type_id",
"name",
"description",
"published",
"group_id"
],
```

And yet, if I look at the field, I see that all and only these fields are nullable, and the rest are required.  For example, 'type_id' is the id of the object I'm fetching, so can never be null, and yet its type is:

```
	// get_universe_types_type_id_type_id
	//
	// type_id integer
	// Required: true
	TypeID *int32 `json:"type_id"`
```

## Swagger specification

https://esi.evetech.net/latest/swagger.json

## Steps to reproduce

1.  Grab the swagger.json file
2.  swagger generate client -f ./swagger.json -A eveapi
3.  Observe that fields that should be required are nullable, and fields that should be nullable are not, in models/get_universe_types_type_id_o_k_body.go

Full output:

```
// Code generated by go-swagger; DO NOT EDIT.

package models

// This file was generated by the swagger tool.
// Editing this file might prove futile when you re-run the swagger generate command

import (
	"strconv"

	strfmt "github.com/go-openapi/strfmt"

	"github.com/go-openapi/errors"
	"github.com/go-openapi/swag"
	"github.com/go-openapi/validate"
)

// GetUniverseTypesTypeIDOKBody get_universe_types_type_id_ok
//
// 200 ok object
// swagger:model getUniverseTypesTypeIdOKBody
type GetUniverseTypesTypeIDOKBody struct {

	// get_universe_types_type_id_capacity
	//
	// capacity number
	Capacity float32 `json:"capacity,omitempty"`

	// get_universe_types_type_id_description
	//
	// description string
	// Required: true
	Description *string `json:"description"`

	// get_universe_types_type_id_dogma_attributes
	//
	// dogma_attributes array
	// Max Items: 1000
	DogmaAttributes []*GetUniverseTypesTypeIDOKBodyDogmaAttributesItems `json:"dogma_attributes"`

	// get_universe_types_type_id_dogma_effects
	//
	// dogma_effects array
	// Max Items: 1000
	DogmaEffects []*GetUniverseTypesTypeIDOKBodyDogmaEffectsItems `json:"dogma_effects"`

	// get_universe_types_type_id_graphic_id
	//
	// graphic_id integer
	GraphicID int32 `json:"graphic_id,omitempty"`

	// get_universe_types_type_id_group_id
	//
	// group_id integer
	// Required: true
	GroupID *int32 `json:"group_id"`

	// get_universe_types_type_id_icon_id
	//
	// icon_id integer
	IconID int32 `json:"icon_id,omitempty"`

	// get_universe_types_type_id_market_group_id
	//
	// This only exists for types that can be put on the market
	MarketGroupID int32 `json:"market_group_id,omitempty"`

	// get_universe_types_type_id_mass
	//
	// mass number
	Mass float32 `json:"mass,omitempty"`

	// get_universe_types_type_id_name
	//
	// name string
	// Required: true
	Name *string `json:"name"`

	// get_universe_types_type_id_packaged_volume
	//
	// packaged_volume number
	PackagedVolume float32 `json:"packaged_volume,omitempty"`

	// get_universe_types_type_id_portion_size
	//
	// portion_size integer
	PortionSize int32 `json:"portion_size,omitempty"`

	// get_universe_types_type_id_published
	//
	// published boolean
	// Required: true
	Published *bool `json:"published"`

	// get_universe_types_type_id_radius
	//
	// radius number
	Radius float32 `json:"radius,omitempty"`

	// get_universe_types_type_id_type_id
	//
	// type_id integer
	// Required: true
	TypeID *int32 `json:"type_id"`

	// get_universe_types_type_id_volume
	//
	// volume number
	Volume float32 `json:"volume,omitempty"`
}

// Validate validates this get universe types type Id o k body
func (m *GetUniverseTypesTypeIDOKBody) Validate(formats strfmt.Registry) error {
	var res []error

	if err := m.validateDescription(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateDogmaAttributes(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateDogmaEffects(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateGroupID(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateName(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validatePublished(formats); err != nil {
		res = append(res, err)
	}

	if err := m.validateTypeID(formats); err != nil {
		res = append(res, err)
	}

	if len(res) > 0 {
		return errors.CompositeValidationError(res...)
	}
	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateDescription(formats strfmt.Registry) error {

	if err := validate.Required("description", "body", m.Description); err != nil {
		return err
	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateDogmaAttributes(formats strfmt.Registry) error {

	if swag.IsZero(m.DogmaAttributes) { // not required
		return nil
	}

	iDogmaAttributesSize := int64(len(m.DogmaAttributes))

	if err := validate.MaxItems("dogma_attributes", "body", iDogmaAttributesSize, 1000); err != nil {
		return err
	}

	for i := 0; i < len(m.DogmaAttributes); i++ {
		if swag.IsZero(m.DogmaAttributes[i]) { // not required
			continue
		}

		if m.DogmaAttributes[i] != nil {
			if err := m.DogmaAttributes[i].Validate(formats); err != nil {
				if ve, ok := err.(*errors.Validation); ok {
					return ve.ValidateName("dogma_attributes" + "." + strconv.Itoa(i))
				}
				return err
			}
		}

	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateDogmaEffects(formats strfmt.Registry) error {

	if swag.IsZero(m.DogmaEffects) { // not required
		return nil
	}

	iDogmaEffectsSize := int64(len(m.DogmaEffects))

	if err := validate.MaxItems("dogma_effects", "body", iDogmaEffectsSize, 1000); err != nil {
		return err
	}

	for i := 0; i < len(m.DogmaEffects); i++ {
		if swag.IsZero(m.DogmaEffects[i]) { // not required
			continue
		}

		if m.DogmaEffects[i] != nil {
			if err := m.DogmaEffects[i].Validate(formats); err != nil {
				if ve, ok := err.(*errors.Validation); ok {
					return ve.ValidateName("dogma_effects" + "." + strconv.Itoa(i))
				}
				return err
			}
		}

	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateGroupID(formats strfmt.Registry) error {

	if err := validate.Required("group_id", "body", m.GroupID); err != nil {
		return err
	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateName(formats strfmt.Registry) error {

	if err := validate.Required("name", "body", m.Name); err != nil {
		return err
	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validatePublished(formats strfmt.Registry) error {

	if err := validate.Required("published", "body", m.Published); err != nil {
		return err
	}

	return nil
}

func (m *GetUniverseTypesTypeIDOKBody) validateTypeID(formats strfmt.Registry) error {

	if err := validate.Required("type_id", "body", m.TypeID); err != nil {
		return err
	}

	return nil
}

// MarshalBinary interface implementation
func (m *GetUniverseTypesTypeIDOKBody) MarshalBinary() ([]byte, error) {
	if m == nil {
		return nil, nil
	}
	return swag.WriteJSON(m)
}

// UnmarshalBinary interface implementation
func (m *GetUniverseTypesTypeIDOKBody) UnmarshalBinary(b []byte) error {
	var res GetUniverseTypesTypeIDOKBody
	if err := swag.ReadJSON(b, &res); err != nil {
		return err
	}
	*m = res
	return nil
}

```

## Environment
swagger version: dev (latest, tested today)
go version: 1.10.2
OS:   Linux 4.16.7-1-ARCH

**Comment by saward** [comment]
I don't really understand, so perhaps I'll ask a question, the answer to which may help enlighten me.  I'm using this as a client to fetch data from the API.

Capacity is, according to the API spec, an optional field.  When I fetch this object from the API, how do I distinguish between capacity being 0 vs null?  It is a float value here, not a pointer, so cannot be null.

This is the payload I am given when I call the API to fetch an object.

**Comment by gregmarr** [comment]
You would need to add `x-nullable: true` to that field of the schema if you need to know that.

**Comment by saward** [comment]
Is there a simple way to specify isnullable for every field across the spec?  Or to provide a customer additional file to supplement or override the spec in parts?

The spec is provided by a third party, and it would be very cumbersome to modify their spec every time it changes (which is, from what I can gather from the docs, what I'd need to do).  I'm hoping therefore there's a way to specify isnullable in a way I can keep static as they continually modify their spec.

Thanks!

**Comment by gregmarr** [comment]
Yeah, having to edit a third party spec is a pain. A command line option for this could probably be added. I have no idea how involved that would be. Maybe a way to set the system-wide default.

**Comment by saward** [comment]
Nor do I have any idea how involved it would be :)

Just to summarise the situation again.  I'm using this to generate a client.  The spec claims that some fields are required, and other optional.  To handle this, normally you might have optional fields as pointers and others not.  However, the generated code produces the reverse of this, for validation reasons from what I can gather.

Nullable variables for optional fields strikes me as something that's required: to tell, for example, whether an integer is actually 0 or not provided at all.

And, as just stated above, modifying the client spec each time it updates is not going to be an easy task.

**Comment by fredbi** [comment]
@saward : this issue with representing fields as pointers or not is a recurring one. 
You may learn more about the whereabouts by searching our issues tagged "nullable".

**Comment by fredbi** [comment]
TL;DR:
-- Required fields generate pointers
- Non-required fields do not
- As for current, we don't make a distinction between an absent optional property and this property set to its type's zero value

**Comment by casualjim** [comment]
The logic is optional fields always have a default of some sort. 
Required fields is actually where you care about the difference between not provided or zero ("", 0, false,...)
